File: pininputlayout/src/main/java/com/hendraanggrian/widget/PinInputLayout.java
package com.hendraanggrian.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.widget.LinearLayout;

import com.hendraanggrian.widget.pininputlayout.PinInput;
import com.hendraanggrian.widget.pininputlayout.R;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class PinInputLayout extends LinearLayout implements TextWatcher, View.OnFocusChangeListener {

    private static final int DEFAULT_PIN_DIGITS = 4;

    @NonNull private final List<PinInput> pinInputs;
    private int focusedPin;
    private State currentState = State.INCOMPLETE;
    @Nullable private OnStateChangedListener onStateChangedListener;
    @Nullable private OnPinChangedListener onPinChangedListener;

    public PinInputLayout(Context context) {
        this(context, null);
    }

    public PinInputLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PinInputLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setOrientation(HORIZONTAL);
        TypedArray array = context.getTheme().obtainStyledAttributes(attrs, R.styleable.PinInputLayout, defStyleAttr, 0);
        try {
            int pinDigits = array.getInt(R.styleable.PinInputLayout_pinDigits, DEFAULT_PIN_DIGITS);
            int pinMargin = (int) array.getDimension(R.styleable.PinInputLayout_pinMargin, context.getResources().getDimension(R.dimen.margin_pin));
            int pinBackground = array.getResourceId(R.styleable.PinInputLayout_pinBackground, -1);
            int pinTextAppearance = array.getResourceId(R.styleable.PinInputLayout_pinTextAppearance, android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP
                    ? android.R.style.TextAppearance_Material_Display1
                    : android.R.style.TextAppearance_Large);
            int pinTextColor = array.getColor(R.styleable.PinInputLayout_pinTextColor, -1);
            float pinTextSize = array.getDimension(R.styleable.PinInputLayout_pinTextSize, -1);

            pinInputs = new ArrayList<>();
            for (int i = 0; i < pinDigits; i++) {
                pinInputs.add(new PinInput(context, this));
                pinInputs.get(i).setMargin(i > 0 ? pinMargin : 0, 0, 0, 0);
                pinInputs.get(i).setTextAppearance(context, pinTextAppearance);
                if (pinBackground != -1)
                    pinInputs.get(i).setBackgroundResource(pinBackground);
                if (pinTextColor != -1)
                    pinInputs.get(i).setTextColor(pinTextColor);
                if (pinTextSize != -1)
                    pinInputs.get(i).setTextSize(pinTextSize);
                addView(pinInputs.get(i));
            }
        } finally {
            array.recycle();
        }
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        if (onStateChangedListener != null) {
            if (arePinsFilled() && currentState != State.COMPLETE) {
                currentState = State.COMPLETE;
                onStateChangedListener.onStateChanged(this, State.COMPLETE);
            } else if (!arePinsFilled() && currentState != State.INCOMPLETE) {
                currentState = State.INCOMPLETE;
                onStateChangedListener.onStateChanged(this, State.INCOMPLETE);
            }
        }
        if (onPinChangedListener != null) {
            String[] pins = new String[pinInputs.size()];
            for (int i = 0; i < pinInputs.size(); i++)
                pins[i] = pinInputs.get(i).getText().toString();
            onPinChangedListener.onPinChanged(this, pins);
        }
    }

    @Override
    public void afterTextChanged(Editable s) {
        if (!s.toString().isEmpty() && focusedPin < pinInputs.size() - 1)
            pinInputs.get(focusedPin + 1).focus();
        else if (s.toString().isEmpty() && focusedPin > 0)
            pinInputs.get(focusedPin - 1).focus();
    }

    @Override
    public void onFocusChange(View v, boolean hasFocus) {
        if (hasFocus)
            focusedPin = pinInputs.indexOf((PinInput) v);
    }

    public void addTextChangedListener(TextWatcher textWatcher) {
        for (PinInput pinInput : pinInputs)
            pinInput.addTextChangedListener(textWatcher);
    }

    public void setOnStateChangedListener(@Nullable OnStateChangedListener onStateChangedListener) {
        this.onStateChangedListener = onStateChangedListener;
    }

    @Nullable
    public OnStateChangedListener getOnStateChangedListener() {
        return onStateChangedListener;
    }

    public void setOnPinChangedListener(@Nullable OnPinChangedListener onPinChangedListener) {
        this.onPinChangedListener = onPinChangedListener;
    }

    @Nullable
    public OnPinChangedListener getOnPinChangedListener() {
        return onPinChangedListener;
    }

    public String getText() {
        String code = "";
        for (PinInput PinInput : pinInputs)
            code += PinInput.getText();
        return code;
    }

    public void clear() {
        for (PinInput pinInput : pinInputs)
            pinInput.setText("");
    }

    public void setInputType(int type) {
        for (PinInput pinInput : pinInputs)
            pinInput.setInputType(type);
    }

    private boolean arePinsFilled() {
        for (PinInput pinInput : pinInputs)
            if (pinInput.getText().toString().isEmpty())
                return false;
        return true;
    }

    public enum State {
        INCOMPLETE, COMPLETE
    }

    public interface OnStateChangedListener {
        void onStateChanged(@NonNull PinInputLayout view, @NonNull State state);
    }

    public interface OnPinChangedListener {
        void onPinChanged(@NonNull PinInputLayout view, @NonNull String... pins);
    }
}

File: pininputlayout/src/com/hendraanggrian/widget/PinInputLayout.kt
package com.hendraanggrian.widget

import android.content.Context
import android.os.Build
import android.support.annotation.AnyRes
import android.support.annotation.ColorInt
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.LinearLayout
import com.hendraanggrian.pininputlayout.R

open class PinInputLayout @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    companion object {
        private const val DEFAULT_PIN_DIGITS = 4
    }

    private val mTextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            if (onStateChangedListener != null) {
                if (areInputsFilled && !mComplete) {
                    mComplete = true
                    onStateChangedListener!!(this@PinInputLayout, true)
                } else if (!areInputsFilled && mComplete) {
                    mComplete = false
                    onStateChangedListener!!(this@PinInputLayout, false)
                }
            }
            if (onPinChangedListener != null) {
                val pins = Array(inputs.size) { "" }
                for (i in inputs.indices) {
                    pins[i] = inputs[i].text.toString()
                }
                onPinChangedListener!!(this@PinInputLayout, pins)
            }
        }

        override fun afterTextChanged(s: Editable) {
            if (!s.toString().isEmpty() && mFocusedInput < inputs.size - 1) {
                inputs[mFocusedInput + 1].requestFocus()
            } else if (s.toString().isEmpty() && mFocusedInput > 0) {
                inputs[mFocusedInput - 1].requestFocus()
            }
        }
    }
    private val mOnFocusedChangeListener: View.OnFocusChangeListener = OnFocusChangeListener { v, hasFocus ->
        if (hasFocus) {
            mFocusedInput = inputs.indexOf(v as PinInput)
        }
    }

    val inputs: List<PinInput>
    var onStateChangedListener: ((PinInputLayout, isComplete: Boolean) -> Unit)? = null
    var onPinChangedListener: ((PinInputLayout, pins: Array<String>) -> Unit)? = null

    private var mFocusedInput: Int = 0
    private var mComplete: Boolean = false
    private var mDigits: Int
    private val mGap: Int
    @AnyRes private var mTextAppearance: Int
    @ColorInt private var mTextColor: Int
    private var mTextSize: Float

    init {
        orientation = LinearLayout.HORIZONTAL

        val a = context.obtainStyledAttributes(attrs, R.styleable.PinInputLayout, defStyleAttr, 0)
        mDigits = a.getInt(R.styleable.PinInputLayout_pinDigits, DEFAULT_PIN_DIGITS)
        mGap = a.getDimensionPixelSize(R.styleable.PinInputLayout_pinGap, context.resources.getDimensionPixelSize(R.dimen.margin_pin))
        mTextAppearance = a.getResourceId(R.styleable.PinInputLayout_pinTextAppearance,
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) android.R.style.TextAppearance_Material_Display1
                else android.R.style.TextAppearance_Large)
        mTextColor = a.getColor(R.styleable.PinInputLayout_pinTextColor, -1)
        mTextSize = a.getDimension(R.styleable.PinInputLayout_pinTextSize, -1f)

        inputs = (0 until mDigits).map {
            val pin = PinInput(context)
            pin.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT).apply { weight = 1f }
            pin.addTextChangedListener(mTextWatcher)
            pin.onFocusChangeListener = mOnFocusedChangeListener
            (pin.layoutParams as LinearLayout.LayoutParams).setMargins(if (it > 0) mGap else 0, 0, 0, 0)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                pin.setTextAppearance(mTextAppearance)
            } else @Suppress("DEPRECATION") {
                pin.setTextAppearance(context, mTextAppearance)
            }
            if (mTextColor != -1) {
                pin.setTextColor(mTextColor)
            }
            if (mTextSize != -1f) {
                pin.textSize = mTextSize
            }
            addView(pin)
            pin
        }
        a.recycle()
    }

    val text: String
        get() {
            var pin = ""
            forEachInput { pin += text }
            return pin
        }

    fun clear() = forEachInput { setText("") }

    private fun forEachInput(block: PinInput.() -> Unit) = inputs.forEach { block(it) }

    private val areInputsFilled: Boolean get() = inputs.none { it.text.toString().isEmpty() }
}